File: day2.nim
type RPS = enum Rock, Paper, Scissors
type Game = tuple[them: RPS, us: RPS]

var games1: seq[Game]
var games2: seq[Game]
var input = open("in/input2")
var line: string
while input.readLine(line):
    if line == "": continue
    var them: RPS
    var us1: RPS
    var us2: RPS
    case line[0]
    of 'A':
        them = Rock
    of 'B':
        them = Paper
    of 'C':
        them = Scissors
    else:
        discard
    case line[2]
    of 'X':
        us1 = Rock
        case them:
        of Rock:
            us2 = Scissors
        of Paper:
            us2 = Rock
        of Scissors:
            us2 = Paper
    of 'Y':
        us1 = Paper
        us2 = them
    of 'Z':
        us1 = Scissors
        case them:
        of Rock:
            us2 = Paper
        of Paper:
            us2 = Scissors
        of Scissors:
            us2 = Rock
    else:
        discard

    games1.add((them, us1))
    games2.add((them, us2))

proc score(game: Game): int =
    case game.us:
    of Rock:
        result += 1
        case game.them:
        of Rock:
            result += 3
        of Paper:
            result += 0
        of Scissors:
            result += 6
    of Paper:
        result += 2
        case game.them:
        of Rock:
            result += 6
        of Paper:
            result += 3
        of Scissors:
            result += 0        
    of Scissors:
        result += 3
        case game.them:
        of Rock:
            result += 0
        of Paper:
            result += 6
        of Scissors:
            result += 3

var total_score = 0
for game in games1:
    total_score += game.score()

echo "Part 1: ", total_score

total_score = 0
for game in games2:
    total_score += game.score()

echo "Part 2: ", total_score

File: zig/day2.zig
const std = @import("std");

const RPS = enum(u8) { Rock = 1, Paper, Scissors };
const Game = struct {
    them: RPS,
    us: RPS,
    fn score(self: *const Game) u32 {
        const scores = [_]u32{ 3, 6, 0 };
        return @enumToInt(self.us) + scores[(3 + @enumToInt(self.us) - @enumToInt(self.them)) % 3];
    }
};

pub fn main() !void {
    const input = @embedFile("in/input2");
    var alloc = std.heap.GeneralPurposeAllocator(.{}){};
    var games1 = std.ArrayList(Game).init(alloc.allocator());
    defer games1.deinit();
    var games2 = std.ArrayList(Game).init(alloc.allocator());
    defer games2.deinit();

    const a = .{
        RPS.Scissors,
        RPS.Rock,
    };
    _ = a;

    var games = std.mem.split(u8, input, "\n");
    while (games.next()) |game| {
        if (game.len == 0) continue;
        const them = switch (game[0]) {
            'A' => RPS.Rock,
            'B' => RPS.Paper,
            'C' => RPS.Scissors,
            else => unreachable,
        };
        const us: struct { RPS, RPS } = switch (game[2]) {
            'X' => .{ RPS.Rock, switch (them) {
                RPS.Rock => RPS.Scissors,
                RPS.Paper => RPS.Rock,
                RPS.Scissors => RPS.Paper,
            } },
            'Y' => .{ RPS.Paper, them },
            'Z' => .{ RPS.Scissors, switch (them) {
                RPS.Rock => RPS.Paper,
                RPS.Paper => RPS.Scissors,
                RPS.Scissors => RPS.Rock,
            } },
            else => unreachable,
        };
        const us1 = us.@"0";
        const us2 = us.@"1";
        try games1.append(.{ .us = us1, .them = them });
        try games2.append(.{ .us = us2, .them = them });
    }

    var score1: u32 = 0;
    for (games1.items) |game| {
        score1 += game.score();
    }
    var score2: u32 = 0;
    for (games2.items) |game| {
        score2 += game.score();
    }
    std.debug.print("Part 1: {}\nPart 2: {}", .{ score1, score2 });
}
